### ISSUE TITLE
Performance not improving/getting worse with an increase of threads for a specific query C++

### ISSUE THREAD

**Issue by tav-singh** [duckdb/duckdb#3164]
#### What happens?
If a UDF in the `WHERE` clause of a `SELECT` statment depends on an aggregated value from a CTE, it triggers a nested loop join (blockwise_nl_join) for the join between CTE and a table, instead of a hash join which is triggered when a constant value is provided as an argument.
Furthermore, the query gets slower with an increase in number of threads in case of a nested loop join, but scales with the number of threads when we have a hash join.
#### To Reproduce
Steps to reproduce the behavior. Bonus points if those are only SQL queries.

**repr.sql** 
```
CREATE TABLE Person(PersonId bigint NOT NULL);
CREATE TABLE Person_Subset_l(lid bigint);
CREATE TABLE Person_Subset_r(rid bigint);
CREATE TABLE Person_ids(lid bigint, rid bigint);
CREATE TABLE src_dest(id int, is_variant boolean, v_size bigint, src bigint, dst bigint);
CREATE TABLE t (id bigint);


INSERT INTO Person SELECT range FROM range(0, 11000);
INSERT INTO Person_Subset_l( SELECT person.rowid from Person ORDER BY md5(person.rowid) LIMIT 10000);
INSERT INTO Person_Subset_r( SELECT person.rowid from Person ORDER BY md5(person.rowid) LIMIT 1000);
INSERT INTO Person_ids( SELECT pl.lid, pr.rid from Person_Subset_l pl, Person_Subset_r pr ORDER BY md5(pl.rowid + pr.rowid) );
INSERT INTO Person_ids VALUES (0 , 0);
INSERT INTO t VALUES(2);

INSERT INTO src_dest (id, is_variant, v_size, src, dst) 
WITH cte as (
	SELECT count(p.personid) as vcount FROM Person p
	)
SELECT 0, false, cte.vcount, p.lid, p.rid FROM cte, Person_ids p;

```

**repr_hash_join.benchmark**
````
# name: benchmark/micro/sql-pgq/repr_hash_join.benchmark
# description: MSBFS 
# group: [sql-pgq]

name Rep
group sql-pgq

init
PRAGMA threads=1;

load benchmark/micro/sql-pgq/repr.sql

run
WITH cte as (
		SELECT min(id) AS id from t
	)
SELECT min(src) 
FROM cte, src_dest where pow(src, 2) = cte.id;


result I
NULL
````
**Timings with 1 thread** 
| name   |      run      |  timing |
|----------|:-------------:|------:|
| repr_hash_join.benchmark |  1 | 0.322763|
| repr_hash_join.benchmark |    2  |   0.335948 |
| repr_hash_join.benchmark |    3  |   0.342714 |
| repr_hash_join.benchmark |    4  |   0.336287 |
| repr_hash_join.benchmark |    5  |   0.336883 |

**Timings with 4 threads** 
| name   |      run      |  timing |
|----------|:-------------:|------:|
| repr_hash_join.benchmark |  1 | 0.104731 |
| repr_hash_join.benchmark|    2  |   0.115166 |
| repr_hash_join.benchmark |    3  |  0.146666 |
| repr_hash_join.benchmark |    4  |   0.123792 |
| repr_hash_join.benchmark |    5  |   0.117263 |

**repr_nl.benchmark**
````
# name: benchmark/micro/sql-pgq/repr_nl.benchmark
# description: MSBFS 
# group: [sql-pgq]

name Rep
group sql-pgq

init
PRAGMA threads=1;

load benchmark/micro/sql-pgq/repr.sql

run
WITH cte as (
		SELECT min(id) AS id from t
	)
SELECT min(src) 
FROM cte, src_dest where pow(src, cte.id) = cte.id;


result I
NULL
````
**Timings with 1 thread** 
| name   |      run      |  timing |
|----------|:-------------:|------:|
| repr_nl.benchmark |  1 | 0.262529 |
| repr_nl.benchmark |    2  |   0.301752 |
| repr_nl.benchmark |    3  |   0.328674 |
| repr_nl.benchmark |    4  |   0.289912 |
| repr_nl.benchmark |    5  |   0.321150 |

**Timings with 4 threads** 
| name   |      run      |  timing |
|----------|:-------------:|------:|
| repr_nl.benchmark |  1 | 0.279668 |
| repr_nl.benchmark |    2  |   0.330281 |
| repr_nl.benchmark |    3  |  0.304953 |
| repr_nl.benchmark |    4  |   0.334316 |
| repr_nl.benchmark |    5  |   0.412563 |

In the `repr_nl.benchmark`, the **timings remain similar** even with an increase of threads while in `repr_hash_join.benchmark`, the timings show a **clear decrease** with an increase of threads.  I have tried to keep the query similar to the kleene star transpilation query which uses the CTE result as an argument of a UDF in the `WHERE` clause; which triggers the nested loop join resulting in this scalability issue.

I used benchmark_runner to compare the performance of both the files and stored them under benchmark/micro/sql-pgq diretory 
 
#### Environment (please complete the following information):
 - OS: [e.g. iOS] Ubuntu 20.04
 - DuckDB Version: v0.3.2 
 - DuckDB Client: C++ 
#### Before Submitting

- [x] **Have you tried this on the latest `master` branch?**
* **Python**: `pip install duckdb --upgrade --pre`
* **R**: `install.packages("https://github.com/duckdb/duckdb/releases/download/master-builds/duckdb_r_src.tar.gz", repos = NULL)`
* **Other Platforms**: You can find binaries [here](https://github.com/duckdb/duckdb/releases/tag/master-builds) or compile from source.

- [x] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?**

**Comment by Mytherin** [comment]
In the second query, you are using the `pow` function that uses data from both sides of the join.
```
WITH cte as (
		SELECT min(id) AS id from t
	)
SELECT min(src) 
FROM cte, src_dest where pow(src, cte.id) = cte.id;
```

`cte.id` is from the left side of the join, `src_dest` is from the right side of the join. Because of this, a hash join cannot be used, and a blockwise NL join is used instead, as is visible in the profiler.

The Blockwise NL join has not been particularly well optimized, so I imagine that is the cause of this problem. There are a number of locks in the code there that might cause the poor scaling behavior.

**Comment by szarnyasg** [comment]
Good point, thanks! We managed to work around the problem for our goals by moving the CTE inside a scalar subquery:

```sql
WITH cte as (
		SELECT min(id) AS id from t
	)
SELECT min(src) 
FROM src_dest
WHERE pow(src, (SELECT id FROM cte)) = (SELECT id FROM cte);
```

This results in a hash join and speeds up approximately linearly as the number of threads increases.
